comictaggerlib/ui/talkeruigenerator.py

```diff
@@ -0,0 +1,239 @@
+            url = widget.text().strip()
+
+    if talkers[talker_id].check_api_url(url):
+        QtWidgets.QMessageBox.information(None, "API Key Test", "URL is valid!")
+    else:
+        QtWidgets.QMessageBox.warning(None, "API Key Test", "URL is NOT valid!")
+
+
+def api_key_btn_connect(
+    btn: QtWidgets.QPushButton,
+    talker_id: str,
+    sources_info: dict[str, QtWidgets.QWidget],
+    talkers: dict[str, ComicTalker],
+) -> None:
+    btn.clicked.connect(lambda: call_check_api_key(talker_id, sources_info, talkers))
+
+
+def api_url_btn_connect(
+    btn: QtWidgets.QPushButton,
+    talker_id: str,
+    sources_info: dict[str, QtWidgets.QWidget],
+    talkers: dict[str, ComicTalker],
+) -> None:
+    btn.clicked.connect(lambda: call_check_api_url(talker_id, sources_info, talkers))
+
+
+def generate_checkbox(option: settngs.Setting, layout: QtWidgets.QGridLayout) -> QtWidgets.QCheckBox:
+    widget = QtWidgets.QCheckBox(option.display_name)
+    widget.setToolTip(option.help)
+    layout.addWidget(widget, layout.rowCount(), 0, 1, -1)
+
+    return widget
+
+
+def generate_spinbox(option: settngs.Setting, layout: QtWidgets.QGridLayout) -> QtWidgets.QSpinBox:
+    row = layout.rowCount()
+    lbl = QtWidgets.QLabel(option.display_name)
+    layout.addWidget(lbl, row, 0)
+    widget = QtWidgets.QSpinBox()
+    widget.setRange(0, 9999)
+    widget.setToolTip(option.help)
+    layout.addWidget(widget, row, 1, alignment=QtCore.Qt.AlignLeft)
+
+    return widget
+
+
+def generate_doublespinbox(option: settngs.Setting, layout: QtWidgets.QGridLayout) -> QtWidgets.QDoubleSpinBox:
+    row = layout.rowCount()
+    lbl = QtWidgets.QLabel(option.display_name)
+    layout.addWidget(lbl, row, 0)
+    widget = QtWidgets.QDoubleSpinBox()
+    widget.setRange(0, 9999.99)
+    widget.setToolTip(option.help)
+    layout.addWidget(widget, row, 1, alignment=QtCore.Qt.AlignLeft)
+
+    return widget
+
+
+def generate_textbox(
+    option: settngs.Setting, layout: QtWidgets.QGridLayout
+) -> tuple[QtWidgets.QLineEdit, QtWidgets.QPushButton]:
+    btn = None
+    row = layout.rowCount()
+    lbl = QtWidgets.QLabel(option.display_name)
+    layout.addWidget(lbl, row, 0)
+    widget = QtWidgets.QLineEdit()
+    widget.setObjectName(option.internal_name)
+    widget.setToolTip(option.help)
+    layout.addWidget(widget, row, 1)
+
+    # Special case for api_key, make a test button
+    if option.internal_name.endswith("api_key"):
+        btn = QtWidgets.QPushButton("Test Key")
+        layout.addWidget(btn, row, 2)
+
+    if option.internal_name.endswith("url"):
+        btn = QtWidgets.QPushButton("Test URL")
+        layout.addWidget(btn, row, 2)
+
+    return widget, btn
+
+
+def settings_to_talker_form(sources: dict[str, QtWidgets.QWidget], config: settngs.Config[settngs.Namespace]) -> None:
+    # Set the active talker in sources combo box
+    sources["talker_source"].setCurrentIndex(sources["talker_source"].findData(config[0].talker_source))
```

Comment:
This can be a single call to `setCurrentText`.
> The setter [setCurrentText]() simply calls [setEditText]\() if the combo box is editable. Otherwise, if there is a matching text in the list, [currentIndex] is set to the corresponding index.

[setCurrentText]: https://doc.qt.io/qt-5/qcombobox.html#currentText-prop
[setEditText]: https://doc.qt.io/qt-5/qcombobox.html#setEditText
[currentIndex]: https://doc.qt.io/qt-5/qcombobox.html#currentIndex-prop

Reply:
My preference for the findData is that is will be unique (talker_id). Also will need to pass in the name (Comic Vine) from talkers (not a hardship but still...). Granted there shouldn't be two talkers with the same title but this is more robust. I can change it if you really want though?

Reply:
I thought this was the ID not the name.
Also lets change "talker_source" to "cbxTalker" or "TalkerComboBox" or something similar